src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt

```diff
@@ -0,0 +1,118 @@
+package org.rust.ide.formatter
+
+import com.intellij.formatting.SpacingBuilder
+import com.intellij.psi.codeStyle.CommonCodeStyleSettings
+import com.intellij.psi.tree.IElementType
+import com.intellij.psi.tree.TokenSet
+import org.rust.lang.core.psi.RustCompositeElementTypes.*
+import org.rust.lang.core.psi.RustKeywordTokenType
+import org.rust.lang.core.psi.RustTokenElementTypes.*
+import com.intellij.psi.tree.TokenSet.create as ts
+
+object RustSpacing {
+    fun createSpacingBuilder(commonSettings: CommonCodeStyleSettings,
+                             rustSettings: RustCodeStyleSettings): SpacingBuilder =
+        SpacingBuilder(commonSettings)
+            // rules defined earlier have higher priority
+            //== some special operators
+            .applyForEach(SPACE_AFTER_OPS) { after(it).spaceIf(true).before(it).spaceIf(false) }
+            .afterInside(AND, ts(REF_TYPE, SELF_ARGUMENT)).spaces(0)
+            .beforeInside(Q, TRY_EXPR).spaces(0)
+            .afterInside(UNARY_OPS, UNARY_EXPR).spaces(0)
+
+            //== attributes
+            .aroundInside(ts(SHA, EXCL, LBRACK, RBRACK), ATTRS).spaces(0)
+            .aroundInside(ts(LPAREN, RPAREN), META_ITEM).spaces(0)
+
+            //== empty parens
+            .between(LPAREN, RPAREN).spaceIf(false)
+            .between(LBRACK, RBRACK).spaceIf(false)
+            .between(LBRACE, RBRACE).spaceIf(false)
+            .betweenInside(OR, OR, LAMBDA_EXPR).spaceIf(false)
+
+            //== paren delimited lists
+            // withinPairInside does not accept TokenSet as parent node set :(
+            // and we cannot create our own, because RuleCondition stuff is private
+            .afterInside(LPAREN, PAREN_LIST_HOLDERS).spaceIf(false)
+            .beforeInside(RPAREN, PAREN_LIST_HOLDERS).spaceIf(false)
+            .afterInside(LBRACK, BRACK_LIST_HOLDERS).spaceIf(false)
+            .beforeInside(RBRACK, BRACK_LIST_HOLDERS).spaceIf(false)
+            .afterInside(LBRACE, BRACE_LIST_HOLDERS).spaceIf(false)
+            .beforeInside(RBRACE, BRACE_LIST_HOLDERS).spaceIf(false)
+            .afterInside(LT, ANGLE_LIST_HOLDERS).spaceIf(false)
+            .beforeInside(GT, ANGLE_LIST_HOLDERS).spaceIf(false)
+            .aroundInside(OR, PARAMS_LIKE).spaceIf(false)
+
+            //== items
+            .between(PARAMS_LIKE, RET_TYPE).spacing(1, 1, 0, true, 0)
+            .before(WHERE_CLAUSE).spacing(1, 1, 0, true, 0)
+            .applyForEach(BLOCK_LIKE) { before(it).spaces(1) }
+            // FIXME(jajakobyly): nonono
+            .beforeInside(LBRACE, ts(FOREIGN_MOD_ITEM, MOD_ITEM)).spaces(1)
```

Comment:
I believe these guys should also get separate PSI elements for their bodies, but I'll fix this in separate PR after merging this.